hazelcast/src/main/java/com/hazelcast/spi/Operation.java

```diff
@@ -178,37 +180,85 @@ public final Operation setReplicaIndex(int replicaIndex) {
     }
 
     /**
-     * Gets the callId of this Operation.
-     *
-     * The callId is used to associate the invocation of an Operation on a remote system, with the response from the execution
-     * of that operation.
-     *
-     * @return the callId.
+     * Gets the call ID of this operation. The call ID is used to associate the invocation of an operation
+     * on a remote system with the response from the execution of that operation.
+     * <ul>
+     *     <li>Initially the call ID is zero.</li>
+     *     <li>When an Invocation of the operation is created, call ID is assigned a positive value.</li>
+     *     <li>When the invocation ends, the operation is {@link #deactivate()}d, but the call ID is preserved.</li>
+     *     <li>The same operation may be involved in a further invocation (retrying); this will assign a
+     *     new call ID and reactivate the operation.</li>
+     * </ul>
+     * @return the call ID
      */
     public final long getCallId() {
-        return callId;
+        return Math.abs(callId);
+    }
+
+    /**
+     * Tells whether this operation is involved in an ongoing invocation. Such an operation always has a
+     * positive call ID.
+     * @return {@code true} if the operation's invocation is active; {@code false} otherwise
+     */
+    final boolean hasActiveInvocation() {
+        return callId > 0;
+    }
+
+    /**
+     * Marks this operation as "not involved in an ongoing invocation".
+     * @return {@code true} if this call deactivated the operation; {@code false} if it was already inactive
+     */
+    final boolean deactivate() {
+        long c = callId;
+        if (c <= 0) {
+            return false;
+        }
+        if (CALL_ID.compareAndSet(this, c, -c)) {
+            return true;
+        }
+        if (callId > 0) {
```

Comment:
what if callId is set back to 0? or is that impossible? If not, maybe that should fail as well?